base/abstractdict.jl

```diff
@@ -217,6 +217,7 @@ Dict{Int64, Int64} with 3 entries:
 """
 function merge!(d::AbstractDict, others::AbstractDict...)
     for other in others
+        sizehint!(d, length(d) + length(other))
```

Comment:
Could we just do `sizehint!(d, sum(length, other; init=0))` once upfront? Also, do we generally assume that `AbstractDict`s define `length`? Otherwise we'd need to check `Base.IteratorSize` frst.

Reply:
My first idea was also to use `sum`. However, merging N Dicts would allocate (N-1)-times more memory than necessary in the worst case when keys are not unique. (This PR only 1-times more).

Not sure, if `AbstractDict` has `length`. Didn't find anything in docs. So, I've added a check inspired by:
https://github.com/JuliaLang/julia/blob/a761cc0b82f88838eb31b29e7387cea35a0cab30/base/abstractset.jl#L102

Reply:
Ah, that makes sense!